File: logging_utility.py
"""
Comprehensive Error Logging Utility for Titan Nova
Provides structured logging, error tracking, and audit trail
"""

import json
import os
from datetime import datetime
from pathlib import Path
from enum import Enum
import traceback
import sys

class LogLevel(Enum):
    DEBUG = "DEBUG"
    INFO = "INFO"
    WARNING = "WARNING"
    ERROR = "ERROR"
    CRITICAL = "CRITICAL"

class TitanLogger:
    """Central logging system for flask_app.py"""
    
    LOG_DIR = Path(os.environ.get("LOG_DIR", "./logs"))
    MAX_LOG_SIZE = 10 * 1024 * 1024  # 10MB
    LOG_RETENTION_DAYS = 30
    
    def __init__(self, name="titan-nova"):
        self.name = name
        self.log_dir = self.LOG_DIR
        self.log_dir.mkdir(parents=True, exist_ok=True)
        self.log_file = self.log_dir / f"{name}.log"
        self.error_log_file = self.log_dir / f"{name}-errors.log"
        self.audit_log_file = self.log_dir / f"{name}-audit.log"
    
    def _get_timestamp(self):
        """Get ISO format timestamp"""
        return datetime.utcnow().isoformat() + "Z"
    
    def _format_log_entry(self, level, message, context=None, error=None):
        """Format structured log entry"""
        entry = {
            "timestamp": self._get_timestamp(),
            "level": level.value,
            "message": message,
            "app": self.name
        }
        
        if context:
            entry["context"] = context
        
        if error:
            entry["error"] = {
                "type": type(error).__name__,
                "message": str(error),
                "traceback": traceback.format_exc()
            }
        
        return json.dumps(entry)
    
    def _check_log_rotation(self, log_file):
        """Rotate log if size exceeds limit"""
        if log_file.exists() and log_file.stat().st_size > self.MAX_LOG_SIZE:
            timestamp = datetime.utcnow().strftime("%Y%m%d_%H%M%S")
            backup = log_file.parent / f"{log_file.stem}_{timestamp}.log"
            log_file.rename(backup)
            print(f"📋 Log rotated: {backup}")
    
    def _write_log(self, log_file, entry):
        """Write log entry to file"""
        try:
            self._check_log_rotation(log_file)
            with open(log_file, "a", encoding="utf-8") as f:
                f.write(entry + "\n")
        except Exception as e:
            print(f"❌ Log write failed: {e}", file=sys.stderr)
    
    def debug(self, message, context=None):
        """Debug level logging"""
        entry = self._format_log_entry(LogLevel.DEBUG, message, context)
        self._write_log(self.log_file, entry)
        if os.environ.get("DEBUG") == "1":
            print(f"🔍 {message}", flush=True)
    
    def info(self, message, context=None):
        """Info level logging"""
        entry = self._format_log_entry(LogLevel.INFO, message, context)
        self._write_log(self.log_file, entry)
        print(f"ℹ️ {message}", flush=True)
    
    def warning(self, message, context=None):
        """Warning level logging"""
        entry = self._format_log_entry(LogLevel.WARNING, message, context)
        self._write_log(self.log_file, entry)
        print(f"⚠️ {message}", flush=True)
    
    def error(self, message, error=None, context=None):
        """Error level logging - always writes to errors.log"""
        entry = self._format_log_entry(LogLevel.ERROR, message, context, error)
        self._write_log(self.log_file, entry)
        self._write_log(self.error_log_file, entry)
        print(f"❌ {message}", flush=True)
        if error:
            print(f"   Exception: {type(error).__name__}: {error}", flush=True)
    
    def critical(self, message, error=None, context=None):
        """Critical level logging - alerts"""
        entry = self._format_log_entry(LogLevel.CRITICAL, message, context, error)
        self._write_log(self.log_file, entry)
        self._write_log(self.error_log_file, entry)
        print(f"🚨 CRITICAL: {message}", flush=True)
    
    def audit(self, action, user_id=None, detail=None, status="success"):
        """Audit trail logging for compliance"""
        entry = {
            "timestamp": self._get_timestamp(),
            "action": action,
            "user_id": user_id,
            "status": status,
            "detail": detail,
            "app": self.name
        }
        try:
            self._check_log_rotation(self.audit_log_file)
            with open(self.audit_log_file, "a", encoding="utf-8") as f:
                f.write(json.dumps(entry) + "\n")
        except Exception as e:
            self.error("Audit log write failed", e)
    
    def log_api_call(self, endpoint, method, status_code, duration_ms, user_id=None, error=None):
        """Log API calls with performance metrics"""
        context = {
            "endpoint": endpoint,
            "method": method,
            "status_code": status_code,
            "duration_ms": duration_ms,
            "user_id": user_id
        }
        
        if status_code >= 500:
            self.error(f"API {method} {endpoint} failed with {status_code}", error, context)
        elif status_code >= 400:
            self.warning(f"API {method} {endpoint} returned {status_code}", context)
        else:
            if duration_ms > 1000:
                self.warning(f"API {method} {endpoint} slow: {duration_ms}ms", context)
            else:
                self.debug(f"API {method} {endpoint} {status_code}", context)
    
    def log_firebase_operation(self, operation, duration_ms, success, error=None):
        """Log Firebase operations"""
        context = {
            "operation": operation,
            "duration_ms": duration_ms,
            "success": success
        }
        
        if not success:
            self.error(f"Firebase {operation} failed", error, context)
        elif duration_ms > 5000:
            self.warning(f"Firebase {operation} slow: {duration_ms}ms", context)
        else:
            self.debug(f"Firebase {operation} OK in {duration_ms}ms", context)
    
    def get_recent_errors(self, limit=10):
        """Retrieve recent errors from log"""
        try:
            with open(self.error_log_file, "r", encoding="utf-8") as f:
                lines = f.readlines()
                return [json.loads(line) for line in lines[-limit:]]
        except Exception as e:
            self.error("Failed to read error log", e)
            return []
    
    def get_audit_trail(self, limit=50, action=None, user_id=None):
        """Retrieve audit trail"""
        try:
            with open(self.audit_log_file, "r", encoding="utf-8") as f:
                lines = f.readlines()
                entries = [json.loads(line) for line in lines[-limit:]]
                
                if action:
                    entries = [e for e in entries if e.get("action") == action]
                if user_id:
                    entries = [e for e in entries if e.get("user_id") == user_id]
                
                return entries
        except Exception as e:
            self.error("Failed to read audit log", e)
            return []


# Global logger instance
logger = TitanLogger("flask_app")


# Usage examples (add to flask_app.py):
"""
from logging_utility import logger

# In API endpoints:
@app.route('/api/setup/load')
def api_setup_load():
    start_time = time.time()
    try:
        # Your code here
        duration = (time.time() - start_time) * 1000
        logger.log_api_call('/api/setup/load', 'GET', 200, duration)
    except Exception as e:
        duration = (time.time() - start_time) * 1000
        logger.log_api_call('/api/setup/load', 'GET', 500, duration, error=e)
        return jsonify({'error': str(e)}), 500

# For Firebase operations:
start = time.time()
try:
    result = load_from_firebase()
    duration = (time.time() - start) * 1000
    logger.log_firebase_operation('load_state', duration, True)
except Exception as e:
    duration = (time.time() - start) * 1000
    logger.log_firebase_operation('load_state', duration, False, e)

# For audit trails:
logger.audit('entry_accepted', user_id=uid, detail={'entry_id': entry.id}, status='success')

# For errors:
try:
    process_payment()
except ValueError as e:
    logger.error('Payment processing failed', e, {'user_id': uid, 'amount': amount})
"""

File: logging_utility.js
/**
 * Comprehensive Logging Utility for Gateway.js
 * Provides structured logging, error tracking, and audit trail for Node.js
 */

const fs = require('fs');
const path = require('path');
const { EventEmitter } = require('events');

class LogLevel {
  static DEBUG = 'DEBUG';
  static INFO = 'INFO';
  static WARNING = 'WARNING';
  static ERROR = 'ERROR';
  static CRITICAL = 'CRITICAL';
}

class TitanLogger extends EventEmitter {
  constructor(name = 'gateway') {
    super();
    this.name = name;
    this.logDir = path.join(process.cwd(), process.env.LOG_DIR || './logs');
    this.maxLogSize = 10 * 1024 * 1024; // 10MB
    this.logFile = path.join(this.logDir, `${name}.log`);
    this.errorLogFile = path.join(this.logDir, `${name}-errors.log`);
    this.auditLogFile = path.join(this.logDir, `${name}-audit.log`);
    
    this._ensureLogDir();
  }

  _ensureLogDir() {
    if (!fs.existsSync(this.logDir)) {
      fs.mkdirSync(this.logDir, { recursive: true });
    }
  }

  _getTimestamp() {
    return new Date().toISOString();
  }

  _formatLogEntry(level, message, context = null, error = null) {
    const entry = {
      timestamp: this._getTimestamp(),
      level,
      message,
      app: this.name,
      pid: process.pid
    };

    if (context) {
      entry.context = context;
    }

    if (error) {
      entry.error = {
        type: error.constructor.name,
        message: error.message,
        stack: error.stack
      };
    }

    return JSON.stringify(entry);
  }

  _checkLogRotation(logFile) {
    try {
      if (fs.existsSync(logFile)) {
        const stats = fs.statSync(logFile);
        if (stats.size > this.maxLogSize) {
          const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
          const backup = `${logFile}.${timestamp}`;
          fs.renameSync(logFile, backup);
          console.log(`📋 Log rotated: ${backup}`);
        }
      }
    } catch (e) {
      console.error('Log rotation failed:', e.message);
    }
  }

  _writeLog(logFile, entry) {
    try {
      this._checkLogRotation(logFile);
      fs.appendFileSync(logFile, entry + '\n', 'utf8');
    } catch (e) {
      console.error(`❌ Log write failed to ${logFile}:`, e.message);
    }
  }

  debug(message, context = null) {
    const entry = this._formatLogEntry(LogLevel.DEBUG, message, context);
    this._writeLog(this.logFile, entry);
    if (process.env.DEBUG === '1') {
      console.log(`🔍 [DEBUG] ${message}`, context || '');
    }
  }

  info(message, context = null) {
    const entry = this._formatLogEntry(LogLevel.INFO, message, context);
    this._writeLog(this.logFile, entry);
    console.log(`ℹ️  [INFO] ${message}`, context || '');
  }

  warning(message, context = null) {
    const entry = this._formatLogEntry(LogLevel.WARNING, message, context);
    this._writeLog(this.logFile, entry);
    console.log(`⚠️  [WARN] ${message}`, context || '');
  }

  error(message, errorObj = null, context = null) {
    const entry = this._formatLogEntry(LogLevel.ERROR, message, context, errorObj);
    this._writeLog(this.logFile, entry);
    this._writeLog(this.errorLogFile, entry);
    console.error(`❌ [ERROR] ${message}`);
    if (errorObj) {
      console.error(`   ${errorObj.constructor.name}: ${errorObj.message}`);
    }
    this.emit('error', { message, error: errorObj, context });
  }

  critical(message, errorObj = null, context = null) {
    const entry = this._formatLogEntry(LogLevel.CRITICAL, message, context, errorObj);
    this._writeLog(this.logFile, entry);
    this._writeLog(this.errorLogFile, entry);
    console.error(`🚨 [CRITICAL] ${message}`);
    this.emit('critical', { message, error: errorObj, context });
  }

  audit(action, userId = null, detail = null, status = 'success') {
    const entry = {
      timestamp: this._getTimestamp(),
      action,
      user_id: userId,
      status,
      detail,
      app: this.name,
      pid: process.pid
    };

    try {
      this._checkLogRotation(this.auditLogFile);
      fs.appendFileSync(this.auditLogFile, JSON.stringify(entry) + '\n', 'utf8');
    } catch (e) {
      this.error('Audit log write failed', e);
    }
  }

  logWhatsAppEvent(eventType, detail = null) {
    const context = {
      event: eventType,
      ...detail
    };
    this.info(`WhatsApp: ${eventType}`, context);
    this.audit(`whatsapp_${eventType}`, null, detail);
  }

  logApiCall(endpoint, method, statusCode, durationMs, userId = null, error = null) {
    const context = {
      endpoint,
      method,
      status_code: statusCode,
      duration_ms: durationMs,
      user_id: userId
    };

    if (statusCode >= 500) {
      this.error(`API ${method} ${endpoint} failed with ${statusCode}`, error, context);
    } else if (statusCode >= 400) {
      this.warning(`API ${method} ${endpoint} returned ${statusCode}`, context);
    } else {
      if (durationMs > 1000) {
        this.warning(`API ${method} ${endpoint} slow: ${durationMs}ms`, context);
      } else {
        this.debug(`API ${method} ${endpoint} ${statusCode}`, context);
      }
    }
  }

  logFirebaseOperation(operation, durationMs, success, error = null) {
    const context = {
      operation,
      duration_ms: durationMs,
      success
    };

    if (!success) {
      this.error(`Firebase ${operation} failed`, error, context);
    } else if (durationMs > 5000) {
      this.warning(`Firebase ${operation} slow: ${durationMs}ms`, context);
    } else {
      this.debug(`Firebase ${operation} OK in ${durationMs}ms`, context);
    }
  }

  logMessageProcessing(messageId, chatJid, type, success, detail = null) {
    const context = {
      message_id: messageId,
      chat_jid: chatJid,
      type,
      ...detail
    };

    if (!success) {
      this.error(`Message processing failed: ${type}`, null, context);
    } else {
      this.debug(`Message processed: ${type}`, context);
    }
    this.audit(`message_${type}`, null, detail, success ? 'success' : 'failed');
  }

  logEntryAccepted(entryId, userId, market, gameType, total) {
    const detail = { entry_id: entryId, user_id: userId, market, game_type: gameType, total };
    this.info(`✅ Entry accepted: ${entryId}`, detail);
    this.audit('entry_accepted', userId, detail);
  }

  logResultSettlement(date, market, stage, result, hitCount, missCount, marketProfit) {
    const detail = {
      date,
      market,
      stage,
      result,
      hit_count: hitCount,
      miss_count: missCount,
      market_profit: marketProfit
    };
    this.info(`🏆 Settlement: ${market} ${stage} ${result}`, detail);
    this.audit('result_settlement', null, detail);
  }

  logSpamGuardAction(chatJid, senderJid, kind, action, count) {
    const detail = {
      chat_jid: chatJid,
      sender_jid: senderJid,
      kind,
      action,
      count
    };
    this.info(`🛡️ SpamGuard ${action}: ${kind} (${count})`, detail);
    this.audit(`spamguard_${action}`, senderJid, detail);
  }

  getRecentErrors(limit = 10) {
    try {
      const data = fs.readFileSync(this.errorLogFile, 'utf8');
      const lines = data.split('\n').filter(Boolean);
      return lines.slice(-limit).map(line => {
        try {
          return JSON.parse(line);
        } catch {
          return { raw: line };
        }
      });
    } catch (e) {
      this.error('Failed to read error log', e);
      return [];
    }
  }

  getAuditTrail(limit = 50, action = null, userId = null) {
    try {
      const data = fs.readFileSync(this.auditLogFile, 'utf8');
      let entries = data.split('\n').filter(Boolean).map(line => {
        try {
          return JSON.parse(line);
        } catch {
          return null;
        }
      }).filter(Boolean);

      if (action) {
        entries = entries.filter(e => e.action === action);
      }
      if (userId) {
        entries = entries.filter(e => e.user_id === userId);
      }

      return entries.slice(-limit);
    } catch (e) {
      this.error('Failed to read audit log', e);
      return [];
    }
  }

  getHealthStatus() {
    return {
      app: this.name,
      timestamp: this._getTimestamp(),
      logFiles: {
        main: fs.existsSync(this.logFile),
        errors: fs.existsSync(this.errorLogFile),
        audit: fs.existsSync(this.auditLogFile)
      },
      logSizes: {
        main: fs.existsSync(this.logFile) ? fs.statSync(this.logFile).size : 0,
        errors: fs.existsSync(this.errorLogFile) ? fs.statSync(this.errorLogFile).size : 0,
        audit: fs.existsSync(this.auditLogFile) ? fs.statSync(this.auditLogFile).size : 0
      }
    };
  }
}

// Global logger instance
const logger = new TitanLogger('gateway');

module.exports = { logger, LogLevel, TitanLogger };

// Usage examples (add to Gateway.js):
/*
const { logger } = require('./logging_utility');

// In WhatsApp connection handler:
sock.ev.on('connection.update', (update) => {
  const { connection, qr } = update;
  if (qr) {
    logger.logWhatsAppEvent('qr_generated', { qr_length: qr.length });
  }
  if (connection === 'open') {
    logger.logWhatsAppEvent('connected');
  }
});

// In entry processing:
const startTime = Date.now();
try {
  const saved = await saveAcceptedEntryToFirebase(parsed, meta);
  if (saved.ok) {
    const duration = Date.now() - startTime;
    logger.logEntryAccepted(saved.entry.id, saved.entry.userId, 
      saved.entry.market, saved.entry.gameType, saved.entry.total);
  }
} catch (e) {
  logger.error('Entry save failed', e, { market: parsed.market });
}

// In API handlers:
app.get('/api/status', async (req, res) => {
  const startTime = Date.now();
  try {
    const state = await fetchFirebaseState();
    const duration = Date.now() - startTime;
    logger.logApiCall('/api/status', 'GET', 200, duration);
    res.json({ status: 'success' });
  } catch (e) {
    const duration = Date.now() - startTime;
    logger.logApiCall('/api/status', 'GET', 500, duration, null, e);
    res.status(500).json({ error: e.message });
  }
});

// In Firebase operations:
const fbStartTime = Date.now();
try {
  const state = await fetchFirebaseState();
  const duration = Date.now() - fbStartTime;
  logger.logFirebaseOperation('fetch_state', duration, true);
} catch (e) {
  const duration = Date.now() - fbStartTime;
  logger.logFirebaseOperation('fetch_state', duration, false, e);
}

// For spam guard:
logger.logSpamGuardAction(chatJid, senderJid, 'link', 'warning', record.count);

// Get health status:
app.get('/logger/health', (req, res) => {
  res.json(logger.getHealthStatus());
});
*/
